### DISCUSSION TITLE
Recovery with DisallowConcurrentExecution does not maintain Trigger order

### DISCUSSION THREAD

**Discussion by EdvardasDlugauskas** [quartznet/quartznet#1121] [Q&A]
Here I will describe the problem we have experienced as well as the temporary workaround we have arrived at. 

The main questions are:
1. Is this in Quartz's scope, is it a Quartz issue, is a Quartz fix in order?
2. Is the workaround sound, is there a better way to achieve the expected result?

### System configuration

**Environment**: clustered (multi-instance)
**Datasource**: SqlServer, AdoJobStore, SqlServerDelegate
**Job type**: plain IJob with DisallowConcurrentExecution, triggered with different parameters from JobDataMap
**Job configuration**: request recovery, store durably
**Trigger**: Job triggered with `await scheduler.TriggerJob(jobKey, jobDataMap)`

### The problem

We have observed that during a recovery (service restart, redeployment) our long running Quartz Job Triggers are sometimes not picked up, instead a different queued (TRIGGER_STATE = "Blocked") Trigger for the Job is set to execute first, ignoring the rule of the ordering of Triggers described in the docs -- [Quartz.Net Priority](https://www.quartz-scheduler.net/documentation/quartz-3.x/tutorial/more-about-triggers.html#priority): "a trigger scheduled to fire at 10:59 will always fire before one scheduled to fire at 11:00".

More specifically, say we have a Quartz Job called SequentialJob with the DisallowConcurrentExecution attribute. We trigger this SequentialJob at 11:00 with trigger A, at 11:01 with trigger B, and at 11:02 with C. The trigger A executes, it's expected to take 10 minutes to finish. At 11:05 the trigger A is executing ("Completed"), B and C are waiting in the queue ("Blocked"). Then, at 11:05, a system redeployment occurs.

**Expected**: the trigger order would remain the same, Quartz will execute the recovery of A first, then B, then C.
**Actual**: the trigger order seems to be set randomly (arbitrarily)

#### Specific case

A specific case from our system. 

I have triggered the SequentialJob 5 times. Here is the QRTZ_TRIGGERS table view:
![Quartz triggers before recovery](https://user-images.githubusercontent.com/21359162/109948170-bc632780-7ce2-11eb-9489-2a8730e32611.png)

I then restarted my local service. This triggered a recovery. The executing trigger was correctly set to RECOVERING_JOBS. However, the NEXT_FIRE_TIME for all triggers is reset (I assume to the recovery time). The trigger order is lost, ordering by NEXT_FIRE_TIME will produce a different order of triggers. The table view:
![Quartz triggers after recovery](https://user-images.githubusercontent.com/21359162/109949028-8ffbdb00-7ce3-11eb-9574-b605a1ab0376.png)

As you can see, the trigger MT_445 was picked to execute instead of the recovering trigger (previously MT_501).

### The workaround

Currently we have the following workaround to mitigate this issue:

For every starting SequentialJob instance, if this is NOT a recovery, perform the following check:
1. Query Quartz to check whether RECOVERY_JOBS triggers exist, if not exists, continue Job execution normally
2. Reschedule the RECOVERY_JOBS trigger with high priority:
```cs
var newRecoveryTrigger = triggerToReschedule.GetTriggerBuilder().Build() as SimpleTriggerImpl;
newRecoveryTrigger.SetNextFireTimeUtc(aLotOfDaysAgo);
newRecoveryTrigger.StartTimeUtc = aLotOfDaysAgo;
newRecoveryTrigger.SetPreviousFireTimeUtc(aLotOfDaysAgo);
newRecoveryTrigger.Priority = aBigNumber;
newRecoveryTrigger.MisfireInstruction = MisfireInstruction.IgnoreMisfirePolicy;
```
3. Reschedule the current running trigger (so we are able to perform step 4, but not lose the current trigger)
4. Return from the SequentialJob without performing any processing, allowing Quartz to pick the next Trigger (the new recovery trigger this time)

This solution seems to solve the issue but requires quite a few extra trips to the Quartz DB. Is there a better way to do this?

### The questions
So, again, the main questions are:

1. Is this in Quartz's scope, is it a Quartz issue, is a Quartz fix in order?
2. Is the workaround sound, is there a better way to achieve the expected result?

**Comment by EdvardasDlugauskas** [comment]
@lahma Sorry for the wall of text ⬆️, but could you confirm that this is a current limitation of Quartz and not us being silly?

Quartz took a huge load off of our shoulders, we're really happy that we chose this package above others. Thank you for your work 👍

**Reply by EdvardasDlugauskas** [reply]
@lahma do you know if this is consistent with the Java version? Or could we contribute to the repo to fix this issue?

**Reply by lahma** [reply]
Thank you for giving should detailed and thorough explanation of the problem, sorry for taking such a long time to respond. I've tried to do some easier things just to make the brain happy. This is not easy 😉 

Looking at the code the scheduled misfire handling should respect the original scheduling:

https://github.com/quartznet/quartznet/blob/34382ba9eba8ca61c7d4a5c3357f15b4cd002f7e/src/Quartz/Impl/AdoJobStore/StdAdoDelegate.cs#L468-L472

So if you have the time I would love the help with this. Logically the original scheduled time should be respected and the misfired job should be fired as soon as possible, cutting in line to get run before the other scheduled ones.

**Reply by EdvardasDlugauskas** [reply]
So say we have A scheduled at 12:00, and B at 12:15. A was blocked until 12:30, and it fired at 12:30 when it got the chance, before B (as it should). During A's execution, a recovery occurs. So now, from what you've shown, it would use the firedTime 12:30 instead of the original 12:00? Would that mean that B at 12:15 is set to run before A which is now using the fire time of 12:30?

This doesn't quite line up with my findings in the screenshots I think, so I'll try to get some time to actually debug the Quartz library code during a recovery to see what exactly is happening in our case...

Thanks for the pointers 👍

**Comment by EdvardasDlugauskas** [comment]
@lahma replying here instead of starting a new discussion, as I think this is related.

I've noticed that when recovering, Quartz first checks misfired triggers and only then recovering triggers:
![image](https://user-images.githubusercontent.com/21359162/165472545-e1eec8d2-220a-4bb6-a509-031c253807c7.png)

I believe it should be the other way around. For example, if we enable DisallowConcurrentExecution and have two triggers: A and B. Say currently A is executing and B is "blocked" - waiting on A to finish. Then on a recovery, we expect continuation of A first, and only after that the recovery of B. This is obviously only relevant for DisallowConcurrentExecution, as otherwise the triggers would run in parallel either way.

I scanned the code trying to understand if there can be any issues if I change it to do recoveries first (swapping 1 and 2). I couldn't find any possible issues but I'm still not sure. I imagine any bugs that might be introduced will be really hard to debug, so we want to be careful.

My question is - do you see any reasons doing recoveries before misfires might not work? Perhaps it was made this way with something in mind? 
If you don't see any possible issues, we'll test it on our branch.

**Reply by EdvardasDlugauskas** [reply]
@lahma bump

I hope this comment is quick for you to answer. Do you think it is safe for us to try swapping code 1 and 2, as in the screenshot?

Thank you for your hard work!

**Reply by EdvardasDlugauskas** [reply]
@lahma bump
